streams/src/main/java/org/apache/kafka/streams/state/internals/buffer/InMemoryTimeOrderedKeyValueBuffer.java

```diff
@@ -89,6 +102,37 @@ public void put(final long time,
         }
     }
 
+    void cleanDrop(final Bytes key) {
+        final TimeKey timeKey = index.remove(key);
+        if (timeKey != null) {
+            sortedMap.remove(timeKey);
+        }
+    }
+
+    private void markDirtyKey(final Bytes key) {
+        dirtyKeys.remove(key);
+        dirtyKeys.add(key);
```

Comment:
so we need to `remove` -- it's a `Set` and a `put` seems to be sufficient to "overwrite" an existing key in the set?

Reply:
It's to impose the desired iteration ordering. When you re-put a key in a LinkedHashSet, it doesn't change its location in the linked list.

In other words, the LHS maintains insertion, not update, order. By removing and adding, we get update order instead.

There's a bigger question about whether this is important or not. I was thinking it would be nice to re-play the events just as they happened. But the buffer keys themselves have a stable total order, which determines the emit ordering, so any flush/replay ordering differences should not be observable.

So I think I'll just change it to a regular set. Let me know if you disagree.

Reply:
I see. Not sure if we need/want this atm -- will introduce some runtime overhead accessing the `Set` twice. I am not opposed, just unsure. \cc @guozhangwang @bbejeck WDYT?

If we keep it as-is, we should add a comment and explain why we do this (otherwise, we might change the later on, because we forget about it).

Reply:
I've removed it already. Problem solved!